// @omnistudio/components/src/modules/c/dataTable/__tests__/dataset.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
export const columns = [
  {
    fieldName: "index",
    label: "index",
    searchable: false,
    sortable: true,
    type: "text"
  },
  {
    editable: "true",
    fieldName: "isActive",
    label: "isActive",
    searchable: false,
    sortable: true,
    type: "text"
  },
  {
    fieldName: "balance",
    label: "balance",
    searchable: false,
    sortable: true,
    type: "text"
  },
  {
    fieldName: "age",
    label: "age",
    searchable: false,
    sortable: true,
    type: "text"
  },
  {
    fieldName: "eyeColor",
    label: "eyeColor",
    searchable: false,
    sortable: true,
    type: "text"
  },
  {
    fieldName: "name",
    label: "name",
    searchable: false,
    sortable: true,
    type: "text"
  },
  {
    fieldName: "gender",
    label: "gender",
    searchable: false,
    sortable: true,
    type: "text"
  },
  {
    fieldName: "company",
    label: "company",
    searchable: false,
    sortable: true,
    type: "text"
  },
  {
    fieldName: "email",
    label: "email",
    searchable: false,
    sortable: true,
    type: "text"
  },
  {
    fieldName: "phone",
    label: "phone",
    searchable: false,
    sortable: true,
    type: "text"
  }
];
export const records = [
  {
    index: 0,
    isActive: false,
    balance: 1256,
    age: 24,
    eyeColor: "green",
    name: "Harriet Noble",
    gender: "female",
    company: "UPDAT",
    email: "harrietnoble@updat.com",
    phone: "+1 (966) 451-2309"
  },
  {
    index: 1,
    isActive: false,
    balance: 2317.43,
    age: 26,
    eyeColor: "brown",
    name: "Delia Bonner",
    gender: "female",
    company: "SOLAREN",
    email: "deliabonner@solaren.com",
    phone: "+1 (970) 554-3357"
  },
  {
    index: 2,
    isActive: false,
    balance: 1373.74,
    age: 39,
    eyeColor: "brown",
    name: "Key Mooney",
    gender: "male",
    company: "VELOS",
    email: "keymooney@velos.com",
    phone: "+1 (948) 417-3771"
  },
  {
    index: 3,
    isActive: true,
    balance: 1623.21,
    age: 25,
    eyeColor: "green",
    name: "Perez Gibbs",
    gender: "male",
    company: "GEOLOGIX",
    email: "perezgibbs@geologix.com",
    phone: "+1 (972) 599-2440"
  },
  {
    index: 4,
    isActive: true,
    balance: 2358.19,
    age: 22,
    eyeColor: "blue",
    name: "Randolph Goodwin",
    gender: "male",
    company: "SNIPS",
    email: "randolphgoodwin@snips.com",
    phone: "+1 (973) 473-2613"
  }
];
