// @omnistudio/cards/src/modules/c/flexCardUtility/flexcardStandardRuntimeXmlObj.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const defaultXmlObjStandardRuntime = {
  targetConfigs: `<targetConfig targets="lightning__AppPage">
                    <property name="debug" type="Boolean"/>
                    <property name="recordId" type="String"/>
                </targetConfig>
                <targetConfig targets="lightning__RecordPage">
                    <property name="debug" type="Boolean"/>
                </targetConfig>`,
  targets: [
    "lightning__RecordPage",
    "lightning__AppPage",
    "lightning__HomePage"
  ],
  apiVersion: 56,
  masterLabel: "",
  description: "",
  runtimeNamespace: "",
  isExposed: true
};

export default defaultXmlObjStandardRuntime;
