// @omnistudio/cards/src/modules/c/flexCardUtils/customLabels.js
/*************************************************************************
 *
 * VLOCITY, INC. CONFIDENTIAL
 * __________________
 *
 *  [2014] - [2020] Vlocity, Inc.
 *  All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Vlocity, Inc. and its suppliers,
 * if any. The intellectual and technical concepts contained
 * herein are proprietary to Vlocity, Inc. and its suppliers and may be
 * covered by U.S. and Foreign Patents, patents in process, and are
 * protected by trade secret or copyright law. Dissemination of this
 * information and reproduction, modification or reverse-engineering
 * of this material, is prohibited unless prior written permission
 * is obtained from Vlocity, Inc.
 *
 * Build: v234.0.0
 */
const requiredPermissionsError =
  "You don't have required permissions to view this card.";
const ErrorToastTitle = "Error";
const flexInvalidIndex = "Flex.index is reserved and can’t be used.";
const multiSelectInput = "Multi Select Input";
const InvalidResponseErrorMessage =
  "Couldn’t complete the action. Change the FlexCard configuration to return an object or array from the action.";
const AppendRecordTypeError =
  "Invalid Value. Only array's are supported as value";
const AppendRecordNoRecordsError =
  "No Records Found . Flexcard should have records to perform Append Records Action";

export {
  requiredPermissionsError,
  ErrorToastTitle,
  flexInvalidIndex,
  multiSelectInput,
  InvalidResponseErrorMessage,
  AppendRecordTypeError,
  AppendRecordNoRecordsError
};
